File: example/learning-agent/java/src/main/java/org/aiddl/example/learning_agent/functions/RandomSimCreator.java
package org.aiddl.example.learning_agent.functions;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

import org.aiddl.core.interfaces.ConfigurableFunction;
import org.aiddl.core.representation.SetTerm;
import org.aiddl.core.representation.Term;
import org.aiddl.common.planning.PlanningTerm;
import org.aiddl.core.function.FunctionRegistry;
import org.aiddl.core.tools.ComboIterator;
import org.aiddl.core.tools.LockableList;
import org.aiddl.core.tools.Logger;

public class RandomSimCreator implements ConfigurableFunction {

	private String name = RandomSimCreator.class.getSimpleName();
	boolean verbose = true;
	
	@Override
	public void configure(Map<Term, Term> settings, FunctionRegistry fReg) {
		this.verbose = Term.sym("true").equals(settings.get(Term.sym("verbose")));
	}

	@Override
	public Term apply( Term args ) {
		Random r = new Random();
		int n = args.getIntValue();
		
		Set<Term> actions = new LinkedHashSet<>();
		
		List<List<Term>> choices = new ArrayList<List<Term>>();
		
		for ( int i = 0 ; i < n ; i++ ) {
			List<Term> stateChoice = new ArrayList<Term>();
			stateChoice.add(Term.sym("true"));
			stateChoice.add(Term.sym("false"));
			choices.add(stateChoice);
		}
		List<Term> state_transitions = new ArrayList<Term>();
		ComboIterator<Term> stateCombos = new ComboIterator<>(choices);
		for ( List<Term> combo : stateCombos ) {
			Set<Term> s_set = new HashSet<>();
			for ( int i = 1 ; i <= combo.size() ; i++ ) {
//				if ( combo.get(i-1).equals(Term.sym("true")) ) {
					s_set.add(
							Term.keyVal(Term.tuple(Term.sym("light"), Term.integer(i)),
									Term.bool(combo.get(i-1).equals(Term.sym("true")))));
//				}
			}
			SetTerm s = Term.set(s_set);
//			Map<Term, Term> s_sva = PlanningTerm.convert2svaMap(s);
			
			if ( verbose ) {
				Logger.msg(name, "State: " + s);
				Logger.incDepth();
			}
			
			Set<Term> transitions_for_s = new LinkedHashSet<>();
			for ( int j = 0 ; j < r.nextInt(3)+1 ; j++ ) {
				Set<Term> effects_set = new LinkedHashSet<Term>();
				Set<Term> s_next = new LinkedHashSet<>();
				
				int numChanges = r.nextInt(2)+1;
				Set<Integer> used = new HashSet<>();
				do {
					used.add(r.nextInt(n)+1);
				} while ( used.size() < numChanges ); 
				
				for ( int i = 1 ; i <= n ; i++ ) {

					Term variable = Term.tuple(Term.sym("light"), Term.integer(i));

					if ( used.contains(i) ) {
//						if ( s.contains(p_i) ) {
////							effects_set.add(Term.tuple(Term.sym("-"), p_i));
//							effects_set.add(Term.keyVal(p_i.getKey(), Term.bool(false)));
//						} else {
//							s_next.add(p_i);
//							effects_set.add(p_i);
//						}
						effects_set.add(Term.keyVal(variable, Term.bool(! s.get(variable).getBooleanValue() )));
						
						s_next.add(Term.keyVal(variable, Term.bool(! s.get(variable).getBooleanValue() )));								
//								Term.keyVal(variable, Term.bool(! s.get(variable).getBooleanValue() )));
					} else {
//						if ( s.contains(p_i) ) {
							s_next.add(Term.keyVal(variable, s.get(variable)));
//						}
					}
				}
				 
				SetTerm effects = Term.set(effects_set);
				
				if ( effects_set.isEmpty() || transitions_for_s.contains(effects) ) {
					j--;
					continue;
				}
//				transitions_for_s.add(effects);
//				Term actionName = Term.tuple(Term.sym("op"), effects);
//				Term actionName = Term.tuple(Term.sym("op"), Term.integer(effects.hashCode()));
				Term actionName = Term.tuple(Term.sym("push"), Term.integer(r.nextInt(n) +1));
				actions.add(actionName);
				state_transitions.add(Term.keyVal(Term.tuple(actionName, s), Term.set(s_next)));
				
				if ( verbose ) Logger.msg(name, "Action: " + actionName);
				if ( verbose ) Logger.msg(name, "Next state: " + s_next);
			}
			Logger.decDepth();
		}
		
		if ( verbose ) {
			Logger.msg(name, "Transitions: "  + state_transitions.size());
			Logger.msg(name, "Actions:"  + actions.size());
		}

		LockableList rList = new LockableList();
		rList.add(Term.keyVal(TransKey, Term.list(state_transitions)));
		rList.add(Term.keyVal(ActionsKey, Term.set(actions)));
		return Term.list(rList);
	}
	
	private static final Term TransKey = Term.sym("state-transitions");
	private static final Term ActionsKey = Term.sym("actions");
}

File: example/learning-agent-scala/src/main/scala/org/aiddl/example/learning_agent_scala/function/RandomSimCreator.scala
package org.aiddl.example.learning_agent_scala.function

import org.aiddl.core.scala.function.{Function, Verbose}
import org.aiddl.core.scala.representation.{Bool, CollectionTerm, KeyVal, ListTerm, SetTerm, Sym, Term, Tuple, Num}
import org.aiddl.common.scala.planning.PlanningTerm
import org.aiddl.core.scala.util.ComboIterator

import scala.util.Random

object RandomSimCreator extends Function with Verbose {

  private val TransKey = Sym("state-transitions")
  private val ActionsKey = Sym("actions")
  override def apply(x: Term): Term = {
    val r = new Random
    val n: Int = x.intoInt

    var actions: Set[Term] = Set.empty
    var choices: List[List[Term]] = Nil

    for (i <- 0 until n) {
      val stateChoice: List[Term] = List(Bool(true), Bool(false))
      choices = stateChoice :: choices
    }
    var state_transitions: List[Term] = Nil
    val stateCombos: ComboIterator[Term] = new ComboIterator[Term](choices)
    for (combo <- stateCombos) {
      var s_set: Set[Term] = Set.empty
      var i: Int = 1
      while (i <= combo.size) {
        s_set = s_set + KeyVal(
          Tuple(Sym("light"), Num(i)),
          Bool(combo(i-1).equals(Bool(true))))
        i += 1
      }
      val s = SetTerm(s_set)

      this.logger.info("State: " + s)
      this.logger.depth += 1
      var transitions_for_s: Set[Term] = Set.empty
      var j: Int = 0
      while (j < r.nextInt(3) + 1) {
        var effects_set: Set[Term] = Set.empty
        var s_next: Set[Term] = Set.empty
        val numChanges: Int = r.nextInt(2) + 1
        var used: Set[Int] = Set.empty
        while (used.size <= numChanges) {
          used = used + (r.nextInt(n) + 1)
        }
        for (i <- 1 to n) {
          val variable = Tuple(Sym("light"), Num(i))
          if (used.contains(i)) {
            effects_set = effects_set + KeyVal(variable, Bool(!s(variable).intoBool.v))
            s_next = s_next + KeyVal(variable, Bool(!s(variable).intoBool.v))
          } else {
            s_next = s_next + KeyVal(variable, s(variable))
          }
        }
        val effects = SetTerm(effects_set)
        if (effects_set.isEmpty || transitions_for_s.contains(effects)) {
          j -= 1
        } else {
          val actionName = Tuple(Sym("push"), Num(r.nextInt(n) + 1))
          actions = actions + actionName
          state_transitions = KeyVal(Tuple(actionName, s), SetTerm(s_next)) :: state_transitions
          this.logger.info("Action: " + actionName)
          this.logger.info("Next state: " + s_next)
          j += 1
        }
      }
      this.logger.depth -= 1
    }

    this.logger.info("Transitions: " + state_transitions.size)
    this.logger.info("Actions:" + actions.size)

    ListTerm(
      KeyVal(TransKey, ListTerm(state_transitions)),
      KeyVal(ActionsKey, SetTerm(actions + Sym("reset")))
    )
  }
}
